### PR TITLE
docs: add KongServiceFacade docs and examples

### LINKED ISSUE #5152: Implement `ServiceFacade` behind an alpha feature gate
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

Implement an abstraction between the Kubernetes `Service` and Kong `Service` (proposed name: `ServiceFacade` - suggestions welcome) that will be usable as a `backendRef` in Gateway API routes and Ingress.

Design doc [link](https://docs.google.com/document/d/1rSdGUMq5zXo340KH_DbaKqDj6_XYuQHbaaw7Z2nV128/edit).

This feature should be opt-in:
- the CRD should not be installed "by default" by the user who comes to KIC
- the support for this feature should be hidden behind an alpha feature gate

### Proposed Solution

1. create a new `ServiceFacade` CRD (distributed separately) similar to the resource below:
    ```yaml
    apiVersion: configuration.konghq.com/v1beta1
    kind: ServiceFacade
    metadata:
      name: payments-service-group1
      annotations:
        konghq.com/plugins: acl-l1
    spec:
      isFacadeFor:
        apiGroup: v1
        kind: Service
        name: payments-service
      port: 8080
    ```
2. Support `ServiceFacade` in rule types in KIC:
    1. Ingress. allow `ServiceFacade` to be set as `defaultBackend` and individual rules' backend
    2. Gateway API: allow all kinds of Gateway API routes to have `ServiceFacade` as a backendRef
    3. Explicitly not supported (to be optionally implemented later if necessary): TCPIngress, UDPIngress
3. Ensure that setting customizations work on ServiceFacade. This includes
    1. Attaching plugins to a `ServiceFacade`
    2. More broadly speaking: attaching other types of settings (including `proxy`/`upstream`) settings

### Additional information

Diagram of the desired use:
image

### Acceptance Criteria

- [ ] The `ServiceFacade` CRD exists and is *not* installed by default
- [ ] The service facade semantics are implemented and only enabled if the service facade alpha feature gate is enabled
- [ ] An installation and usage guide for ServiceFacade exists (including alpha warnings)
- [ ] If enabled, ServiceFacade works as a backend for `Ingress`
- [ ] Plugin attaching to a `ServiceFacade` works
- [ ] All other customizations configurable on a `Service` (annotations) work

### PR DESCRIPTION
**Pull request by czeslavo**
**What this PR does / why we need it**:

Adds a new `FEATURE_GATES.md` section "Using KongServiceFacade" explaining the basics of `KongServiceFacade`. It also references a new example manifest that demonstrates complete usage of `KongServiceFacade` to set a different set of plugins for the same Kubernetes `Service`.

The instructions in the example manifest depend on https://github.com/Kong/charts/pull/963 which adds RBACs for `KongServiceFacade`.

PR also adds an integration test that verifies the new example manifest to make sure it doesn't get outdated.

**Which issue this PR fixes**:

Closes #5152.

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: FEATURE_GATES.md [hunk 1]
```diff
@@ -168,3 +168,89 @@ services: if two HTTPRoute rules use both serviceA and serviceB in their
...
+spec:
+  ingressClassName: kong
+  rules:
+  - http:
+      paths:
+      - path: /my-service
+        pathType: Prefix
+        backend:
+          resource:
+            apiGroup: incubator.konghq.com
```
Comment (mheap) [inline FEATURE_GATES.md:237]: This is a real URL that redirects to docs. Do we have a naming scheme for api versions with KIC/for k8s at Kong in general?

File: examples/kong-service-facade.yaml [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+            backend:
+              # The /alpha path uses the KongServiceFacade secured with key-auth plugin.
+              resource:
+                apiGroup: incubator.konghq.com
+                kind: KongServiceFacade
+                name: svc-facade-alpha
+          - path: /beta
+            pathType: Prefix
+            backend:
+              # The /beta path uses the KongServiceFacade secured with key-auth plugin.
```
Comment (mheap) [inline examples/kong-service-facade.yaml:134]:
```suggestion
              # The /beta path uses the KongServiceFacade secured with basic-auth plugin.
```

File: FEATURE_GATES.md (hunk 1 above)
Comment (czeslavo) [reply FEATURE_GATES.md:237]: > This is a real URL that redirects to docs.
As long as we are in control (Kong) of the domain, URL being real shouldn't be a big deal. No one expects API groups to be valid/invalid URLs but unique.

>Do we have a naming scheme for api versions with KIC/for k8s at Kong in general?
In Gateway Operator we're using `gateway-operator.konghq.com`. In KIC we've been using `configuration.konghq.com` only so far. Kuma uses `kuma.io` domain. 

We cannot fix `configuration.konghq.com` to be KIC-specific enough, but if you like we can make `incubator` to be `incubator.ingresscontroller.konghq.com` so we're sure we're in a KIC namespace.

What I don't like about it is that we'll be inconsistent by doing so.

Comment (czeslavo) [reply FEATURE_GATES.md:237]: After the Slack discussion: https://github.com/Kong/kubernetes-ingress-controller/pull/5302

Comment (czeslavo) [reply FEATURE_GATES.md]: Waiting for [#968](https://github.com/Kong/charts/pull/968), [#969](https://github.com/Kong/charts/pull/969) in charts to be merged to resolve this discussion.

File: FEATURE_GATES.md [hunk 3]
```diff
@@ -168,3 +168,89 @@ services: if two HTTPRoute rules use both serviceA and serviceB in their
 backendRefs, KIC will generate a single Kong service with endpoints from both
 serviceA and serviceB, named for the first rule and match indices with that
 combination of Services.
+
+## Using KongServiceFacade
+
+In KIC 3.1.0 we introduced a new feature called `KongServiceFacade`. Currently, we only
+support `KongServiceFacade` to be used as a backend for `netv1.Ingress`es.
```
Comment (mflendrich) [inline FEATURE_GATES.md:175]:
```suggestion
support `KongServiceFacade` to be used as a backend for `networking.k8s.io/v1` `Ingress`es.
```
`netv1` is jargon that I don't expect to be understandable by the general public :slightly_smiling_face:

File: FEATURE_GATES.md [hunk 4]
```diff
@@ -168,3 +168,89 @@ services: if two HTTPRoute rules use both serviceA and serviceB in their
...
+```
+
+Please note the `KongServiceFacade` must be in the same namespace as the `Ingress` that uses it.
+
+An advantage of using `KongServiceFacade` over plain `corev1.Service`s is that you can create multiple
+`KongServiceFacade`s that point to the same `Service` and KIC will always generate one Kong Service per each
+`KongServiceFacade`. That enables you to, e.g., use different `KongPlugin`s for each `KongServiceFacade`
+while still pointing to the same Kubernetes `Service`. A single `KongServiceFacade` may be used in multiple
+`Ingress`es and customization done through the `KongServiceFacade`'s annotations will be honored in all of them
+on a single Kong Service level (no need to duplicate annotations in multiple `Ingress`es).
```
Comment (mflendrich) [inline FEATURE_GATES.md:250]:
```suggestion
on a single Kong Service level (no need to duplicate annotations in multiple `Ingress`es).

A recommended, generally available alternative to `KongServiceFacade` exists: you can
create several Kubernetes `Service`s with the same selector. `KongServiceFacade` should
be useful if you are unable to use the `Service` alternative. Reasons known at the
moment of writing are: using certain mesh solutions (Consul Connect) that [impose a
"single service only" requirement](https://github.com/hashicorp/consul-k8s/issues/68) and (hypothetical) performance implications of
repetitive kube-proxy reconciliation.
```

**Review by mflendrich — COMMENTED** [review]
Looking good :+1: 

Only the kustomize command comment is blocking (as the original URL did not work for me with a spurious kustomize error).

**Comment by czeslavo** [comment]
@mheap @mflendrich Thanks for your comments, PTAL again 👍

**Review by mheap — APPROVED** [review]

**Review by mflendrich — APPROVED** [review]

**Pull request merged**